File: Java/hellointerview/01-two-pointers/trapping_rain_water.java
class Solution {
  public static int trappingWater(int[] heights) {
    if (heights.length == 0) {
      return 0;
    }
    int left = 0;
    int right = heights.length - 1;
    int leftMax = heights[left];
    int rightMax = heights[right];
    int result = 0;

    while (left < right) {
      if (leftMax < rightMax) {
        left++;
        if (heights[left] > leftMax) {
          leftMax = heights[left];
        } else {
          count += leftMax - heights[left];
        }
      } else {
        right--;
        if (heights[right] >= rightMax) {
          rightMax = heights[right];
        } else {
          count += rightMax - heights[right];
        }
      }
    }
  }
}

File: TS/hellointerview/two-pointers/trappingRainWater.ts

class Solution {
  trappingRainWater(heights: number[]): number {
    if (heights.length === 0) {
      return 0;
    }

    let left = 0;
    let leftMax = heights[left];
    let right = heights.length - 1;
    let rightMax = heights[right];
    let count = 0;

    while (left < right) {

      if (leftMax < rightMax) {
        left++;
        if (heights[left] >= heights[leftMax]) {
          leftMax = heights[left];
        } else {
          count += leftMax - heights[left];
        }
      } else {
        right++;
        if (heights[right] >= heights[rightMax]) {
          rightMax = heights[right];
        } else {
          count += rightMax - heights[right]
        }
      }
    }
    return count;
  }
}
